### PR TITLE
Add a websocket transport

### PR DESCRIPTION
**Pull request by tomaka**
Based over #81 
cc #23 

Adds the `libp2p-websocket` crate that allows using websockets as a transport.

One things to note is that it provides two completely separate implementations: one for desktop platforms, and one for emscripten.

On emscripten, the transport is a stand-alone, and will handle multiaddresses that look like `/ipN/.../tcp/.../ws`. On desktop, the transport needs to be passed an underlying transport (such as a `TcpConfig`), and will only handle the `/ws` suffix.

The purpose of using websockets is to enable nodes that run in the browser. I have a working example in another branch, which I didn't push because it needs a lot of work.

You can already try to connect two nodes together through websockets by manually providing a dialing or listening address to the `echo` examples. For example, run `echo-server /ip4/0.0.0.0/tcp/10333/ws`, and `echo-dialer /ip4/127.0.0.1/tcp/10333/ws`.

Note that if you try to connect a node that expects websockets to a node that doesn't expect websockets, or vice-versa, you will get a "cryptic" error (something like "failed to negotiate handshake"). It is not possible to configure the dialer or listener to try websockets and fall back to not-using-websockets if that fails, but I don't think it is a problem.

### CODE REVIEW

File: libp2p-swarm/src/transport.rs
Comment (Vurich) [inline libp2p-swarm/src/transport.rs]: Why do you forward the `addr` from `first` and `second` and then `.map` it away immediately?

File: libp2p-swarm/src/transport.rs
Comment (Vurich) [inline libp2p-swarm/src/transport.rs]: We can do this using `as fn(_) -> _` and then copying the giant resulting type signature out of the error message, although it's your call as to whether this is worth doing.

File: libp2p-swarm/src/transport.rs
Comment (Vurich) [inline libp2p-swarm/src/transport.rs]: Mixed tabs and spaces

File: libp2p-swarm/src/transport.rs
Comment (Vurich) [inline libp2p-swarm/src/transport.rs]: Same as above, we can do this without `Box`

File: libp2p-tcp-transport/src/lib.rs
Comment (Vurich) [inline libp2p-tcp-transport/src/lib.rs]: Println left in

File: libp2p-websocket/README.md
Comment (Vurich) [inline libp2p-websocket/README.md]: A nicer way to do this would be

````
```rust
# || {
let _ = ws_config.dial(Multiaddr::new("/ip4/40.41.42.43/tcp/12345/ws").unwrap());
# };
```
````

But that doesn't work outside of rustdoc

File: libp2p-websocket/src/browser.rs
Comment (Vurich) [inline libp2p-websocket/src/browser.rs]: What's the reasoning for this name?

File: libp2p-websocket/src/browser.rs
Comment (Vurich) [inline libp2p-websocket/src/browser.rs]: It'd be nice if the parameter of `RwStreamSink` was a type alias

File: libp2p-websocket/src/browser.rs
Comment (Vurich) [inline libp2p-websocket/src/browser.rs]: Here and elsewhere: blocks should start with a space (so `js! { }` instead of `js!{ }`)

File: libp2p-websocket/src/desktop.rs
Comment (Vurich) [inline libp2p-websocket/src/desktop.rs]: Can we get a source for this? Why is `ws` always valid? What if in the future we do validation to prevent `/udp/.../ws`?

File: libp2p-websocket/src/desktop.rs
Comment (Vurich) [inline libp2p-websocket/src/desktop.rs]: Either turn this into a real error with "pings and pongs are unimplemented" as the message or mark this struct as experimental (or something similar), we can't have the server explode if you send a ping.

File: libp2p-websocket/src/desktop.rs
Comment (Vurich) [inline libp2p-websocket/src/desktop.rs]: Mixed tabs and spaces

File: libp2p-websocket/src/desktop.rs
Comment (Vurich) [inline libp2p-websocket/src/desktop.rs]: Can we fork the WS library to make this cleaner or is it not worth it?

File: libp2p-websocket/src/lib.rs
Comment (Vurich) [inline libp2p-websocket/src/lib.rs]: Why `ignore` and not `no_run`?

File: libp2p-websocket/src/browser.rs
Comment (Vurich) [inline libp2p-websocket/src/browser.rs]: I don't like how the signature for this method changes depending on your platform but otherwise everything acts the same, it seems like it's just asking for confusing error messages when you switch platform and expect everything to work the same. Maybe have `BrowserWsConfig` and `WsConfig`?

**Review by eira-fransham — CHANGES_REQUESTED** [review]

File: libp2p-swarm/src/transport.rs
Comment (tomaka) [reply libp2p-swarm/src/transport.rs]: We can't because unfortunately one of the closures captures a value.

File: libp2p-swarm/src/transport.rs
Comment (tomaka) [reply libp2p-swarm/src/transport.rs]: It's not the address that is mapped away but a `SelectNext` object.

File: libp2p-swarm/src/transport.rs
Comment (tomaka) [reply libp2p-swarm/src/transport.rs]: We also move stuff in the closure here :-/

File: libp2p-websocket/src/browser.rs
Comment (tomaka) [reply libp2p-websocket/src/browser.rs]: It is also true that `WsConfig` (the desktop one) could be used from inside nodejs or another similar environment, which makes the name `BrowserWsConfig` more relevant.

File: libp2p-websocket/src/desktop.rs
Comment (tomaka) [reply libp2p-websocket/src/desktop.rs]: The `encapsulate` method parses the parameter, and adds it to the end. An error can only happen here if we fail to parse the string `"/ws"`. I agree that parsing a string every time is bad, so I will change it.

File: libp2p-websocket/src/lib.rs
Comment (tomaka) [reply libp2p-websocket/src/lib.rs]: This demonstrates the emscripten code, but the doctests are compiled for x86 and hence will fail to compile.

File: libp2p-websocket/src/desktop.rs
Comment (tomaka) [reply libp2p-websocket/src/desktop.rs]: I would say it's not worth it. Since it's a breaking change they would have to bump their version, leading to more burden.

**Comment by tomaka** [comment]
Should be good now.

The ping/pong messages now produce an error and/or close the socket instead of panicking.
This seems like a lazy solution, but handling them is really really not easy. An issue should be opened after this PR is merged for that.

**Review by eira-fransham — APPROVED** [review]

**Pull request merged**
